src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs

```diff
@@ -87,8 +81,21 @@ public void InstallWorkloadPack(PackInfo packInfo, SdkFeatureBand sdkFeatureBand
                     {
                         if (!PackIsInstalled(packInfo))
                         {
-                            var packagePath = _nugetPackageDownloader.DownloadPackageAsync(new PackageId(packInfo.ResolvedPackageId), new NuGetVersion(packInfo.Version)).Result;
-                            tempFilesToDelete.Add(packagePath);
+                            string packagePath;
+                            if (string.IsNullOrWhiteSpace(offlineCache))
+                            {
+                                packagePath = _nugetPackageDownloader.DownloadPackageAsync(new PackageId(packInfo.ResolvedPackageId), new NuGetVersion(packInfo.Version)).Result;
+                                tempFilesToDelete.Add(packagePath);
+                            }
+                            else
+                            {
+                                _reporter.WriteLine(string.Format(LocalizableStrings.UsingCacheForPackInstall, packInfo.Id, packInfo.Version, offlineCache));
```

Comment:
Offline cache should have a matching manifest file and we should install according to that. And once the packs are installed, the manfiest file should be installed too

Reply:
Where should that manifest come from? Should we add logic to download the latest manifest package, extract it and determine which packs to download and cache based on that manifest? cc @dsplaisted

Reply:
> Should we add logic to download the latest manifest package, extract it and determine which packs to download and cache based on that manifest

I think so. For a normal install, we download the latest manifest first and use it to determine the installation too right?

Reply:
Yes, although with the normal installation we install the new manifests to the manifest directory before we start using them. You're saying we should read the manifests from the temp directory, not fully install them, right? Or are we okay to fully install them before determining the packs to download?

Reply:
> Yes, although with the normal installation we install the new manifests to the manifest directory before we start using them

Hmm, I don't think that's spec's behavior. However, if we do that first, and the install failed in the middle, we would invalid all installed workload since they mismatch with the manifest and should be garbage collected.

The temp folder is secure so we can do our work base on the temp manifest.

> we'll have to create a new manifest resolver to determine which packs to install

We cannot just point the manifest resolver  to the temp manifests? 

I think we kind of have too. Manifest should be always in sync with the package as part of the installation.

Reply:
But if it is too hard we can still let offline cache to have a manifest package. And the first action of offline install is to install this offline manifest, and the do the rest. If the manifest is older than what's install, stop

For "if we do that first, and the install failed in the middle, we would invalid all installed workload since they mismatch with the manifest and should be garbage collected." This could be not ideal user experience. We could fix that after we ship

Reply:
> However, if we do that first, and the install failed in the middle, we would invalid all installed workload since they mismatch with the manifest and should be garbage collected.

Yes, that's why any manifest updates are part of the rollback, so if install fails we roll back the manifests as well as the packs.

> We cannot just point the manifest resolver to the temp manifests?

I haven't looked at it closely, but right now the manifest resolver takes in the dotnet root directory and appears to have some logic that involves more than just the manifests directory. I'll take a closer look and work on finding a way to install and extract to a temp directory, although it might be big enough that I'll want to add downloading manifests to an offline cache in a separate PR

Reply:
So just update manifest from the offline cache instead of from the internet might work.

Reply:
> So does that mean we'll need a full set of manifests in a temp directory, then we'll have to create a new manifest resolver to determine which packs to install? It seems like a lot of duplicated work and since the manifest reader/ resolver works off of the dotnet root directory right now, we'll need to make significant changes there to enable working from a temp directory.

Yes, this is what I think we should basically do.

It doesn't seem like the code changes will be too bad to me.  We'll want a new implementation of `IWorkloadManifestProvider` instead of `SdkDirectoryWorkloadManifestProvider`, which we can point at a temporary folder where we put the updated manifests.  I think we can use the existing `WorkloadResolver` class.  It only uses the `dotnetRootPath` to locate the runtime identifier chain (which is still what we want to do in the cache creation case), and to determine the path where workload packs should be installed (which we shouldn't need to do if we're just downloading them to a folder).

If the `--skip-manifest-update` option is specified, than rather then getting the latest manifest packages, we should get the versions matching what is currently installed.

When we install workloads using an offline cache, we should first update the advertising manifests using the manifests from the offline cache.  If the latest version of a manifest in the offline cache is older than what is currently installed then it should error out.

Reply:
Sounds good, thanks for the clarification. I'm going to go ahead and add manifest caching as a separate task in https://github.com/dotnet/sdk/issues/16483 and try to get this PR in without it, since it will help unblock VS for mac.